Contributing my new connector `owasp_api_vulns` created for the OWASP API Vulnerability Advisor project submission for the AI Accelerate Hackathon 2025 Fivetran Challenge.

aksaha9 opened the pull request:
### Description of Change
Contributing my new connector `owasp_api_vulns` created for the OWASP API Vulnerability Advisor project submission for the AI Accelerate Hackathon 2025 Fivetran Challenge.

**Key Features**:
- Fetches CVE data from NVD API for OWASP Top 10 API vulnerabilities and XSS (CWE-79 as edge case).
- Upserts to BigQuery, excluding UNKNOWN severity CVEs for high-quality dataset.
- Supports full sync (~39909 rows) and incremental sync (~77 rows daily for CWE-79).
- Configurable CWE list via `configuration.json` for BAU updates.
- Custom logging (`standard`/`debug`) and sync log table (`owasp_api_sync_log`).

**Testing**:
Successfully tested with `fivetran debug --configuration configuration.json`. 

Successfully tested with `fivetran debug --configuration configuration.json`. Results of local run -
Operation       | Calls     
----------------+------------
Upserts         | 39909     
Updates         | 0         
Deletes         | 0         
Truncates       | 0         
SchemaChanges   | 2         
Checkpoints     | 1         
Note: Fivetran debug's performance is limited by your local machine's resources. Your connector will run faster in production.
read about production system resources at https://fivetran.com/docs/connector-sdk/working-with-connector-sdk#systemresources 
 INFO: Fivetran-Tester-Process: Sync SUCCEEDED 

Screenshot of successful syncs from fivetran dashboard:
<img width="1915" height="875" alt="Full sync screenshot" src="https://github.com/user-attachments/assets/eef2d8be-3824-4178-951b-f601d8fcdd5d" />

Screenshot of data pushed into the BigQuery dataset
Sync log fivetran table - <img width="1911" height="769" alt="image" src="https://github.com/user-attachments/assets/acd5c2f0-fee6-4d02-a4d4-cd1d0eaeec45" />
CVE data fivetran table 
<img width="1920" height="1004" alt="image" src="https://github.com/user-attachments/assets/c18b955f-83fc-4f90-b235-24aa469a7a49" />


**Hackathon Context**:
- Devpost submission: https://devpost.com/software/owasp-api-vulnerability-advisor
- Demo video: https://www.youtube.com/watch?v=R-k66rWTsyE&list=LL
- RAG app: https://owasp-advisor-661184527375.us-central1.run.app

**License**: MIT License (as per Fivetran Connector SDK).

### Checklist
- [x] Tested the connector with `fivetran debug` command.
- [x] Added/Updated example specific README.md file.
- [x] Followed Python Coding Standards.

aksaha9 committed:
Code and Files for the new owasp_api_vulns connector for the OWASP API Vulnerability advisor project for the AI Accelerate Hackathon Fivetran challenge

aksaha9 committed:
Merge branch 'main' of https://github.com/aksaha9/fivetran_connector_sdk into main
Updated the forked repo from the parent repo to bring in the changes and now comitting code changes for the new owasp_api_vulns connector for the AI accelerate hackathon fivetran challenge submission owasp-api-vulnerability-advisor project submission by me - Ashish Kumar Saha

aksaha9 commented:
Hi Team, Could you please assign the label - "accel Google hack 2025" This is related to my submission for AI accelerate hackathon https://ai-accelerate.devpost.com/ which I made yesterday - https://devpost.com/software/owasp-api-vulnerability-advisor?ref_content=user-portfolio&ref_feature=in_progress

aksaha9 commented:
Hi @fivetran-sahilkhirwal @ @fivetran-surabhisingh, @fivetran-monicadholwani — could you please review PR #335 (owasp_api_vulns connector)? 

_What to check:_ 

1. connector logic and sync behaviour (BigQuery upsert), 
2. configuration.json and README, 
3. and any CI/packaging concerns. 

_Testing & validation:_

1. I ran fivetran debug locally (passed) — full sync produced ~39,909 rows; 
2. incremental runs ~77 rows/day (CWE-79). 
3. I also deployed the connector to my Fivetran trial account — live syncs are succeeding and data is being pushed to my BigQuery dataset in my GCP project. 
4. Screenshots of the Fivetran dashboard and BigQuery are included in the PR description. 

_Note:_ I don’t have push access, so I can’t formally “request review.” If you or someone on @fivetran/menon or @fivetran/tech-writers can re-request reviewers or add an individual reviewer, that would be very helpful. Thanks!

aksaha9 committed:
Trigger CI

aksaha9 commented:
Hi @fivetran-sahilkhirwal  @varundhall  could you please review my PR https://github.com/fivetran/fivetran_connector_sdk/pull/335 (owasp_api_vulns connector)?
Thanks,
Ashish

fivetran-sahilkhirwal commented on .gitignore:
We should not add gitignore. Please remove this file

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,296 @@
+import datetime
```
https://github.com/fivetran/fivetran_connector_sdk/blob/379d829f03786b2cfaf0bdbdfd0cc16d71a6232e/template_example_connector/connector.py#L1-L22
This is the template for a connector.
Please follow this exact template for the start of the file.
- Add a pydoc about the example with the standard documentation links at the top
- Add the fivetran imports with the comments ( as present in template )
- Add one line comments for the use of each import

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,296 @@
+        logger.warning(f"Date parsing error for {date_str}: {str(e)} - Using original string")
+        return date_str
+
+def update(configuration: dict, state: dict):
```
The template for the standard docstring:
https://github.com/fivetran/fivetran_connector_sdk/blob/379d829f03786b2cfaf0bdbdfd0cc16d71a6232e/template_example_connector/connector.py#L86-L94

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,296 @@
+def update(configuration: dict, state: dict):
+    start_time = time.time()
+    
+    base_url = "https://services.nvd.nist.gov/rest/json/cves/2.0"
```
+1

fivetran-sahilkhirwal reviewed (changes requested):
Please check the template for the examples: https://github.com/fivetran/fivetran_connector_sdk/blob/main/template_example_connector/connector.py
Address the existing comments and also ensure that the PR checks pass :)

aksaha9 committed:
Revert .gitignore to fivetran official upstream/main to address PR review feedback

aksaha9 committed:
changes to adress PR feedback

aksaha9 committed:
Update connectors/owasp_api_vulns/configuration.json.example

Co-authored-by: Sahil Khirwal <sahil.khirwal@fivetran.com>

aksaha9 committed:
changes to adress PR feedback

aksaha9 committed:
Merge branch 'feature/owasp_api_vulns_connector' of https://github.com/aksaha9/fivetran_connector_sdk into feature/owasp_api_vulns_connector
Replacing actual values with placeholders as advised

aksaha9 committed:
changes to adress PR feedback

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/README.md:
```diff
@@ -0,0 +1,111 @@
+- Operating system:
+  - Windows: 10 or later (64-bit only)
+  - macOS: 13 (Ventura) or later (Apple Silicon [arm64] or Intel [x86_64])
+  - Linux: Distributions such as Ubuntu 20.04 or later, Debian 10 or later, or Amazon Linux 2 or later (arm64 or x86_64)
```
+1

aksaha9 replied:
Changes made to address all feedback

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/README.md:
```diff
@@ -0,0 +1,111 @@
+## Authentication
+This connector authenticates with the NVD 2.0 API using an API key. The key is provided in the `configuration.json` file and sent in the `apiKey` header of each request.
+
+To obtain an API key, you must register on the NVD website.
```
+1

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,341 @@
+from fivetran_connector_sdk import Logging as log  # For enabling Logs in the connector code
+from fivetran_connector_sdk import Operations as op  # For performing data operations like upsert and checkpoint
+
+""" SOURCE-SPECIFIC IMPORTS """
```
```suggestion
```

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,341 @@
+__RESULTS_PER_PAGE = 2000
+__API_RATE_LIMIT_DELAY = 0.6
+
+def schema(configuration: dict):
```
+1

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,341 @@
+        return None
+
+    if is_debug_logging:
+        log.info(f"Processing CVE {cve_id} with CWEs: {cve_cwes}, Description: {desc_lower[:50]}...")
```
+1

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,341 @@
+    upsert_count_per_cwe = 0
+    
+    while True:
+        response = requests.get(base_url, params=params, headers=headers)
```
+1

fivetran-sahilkhirwal replied:
We should also extract the logic of making api call and retries in a separate function for better maintainability :)

fivetran-sahilkhirwal commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,341 @@
+        "sync_type": sync_type,
+        "total_rows_upserted": total_upserts
+    })
+    op.checkpoint({"last_sync_time": sync_datetime})
```
+1

fivetran-sahilkhirwal reviewed (commented):
Please address these comments

Also please check the PR check failure for flake8 and black issue

aksaha9 commented:
Hi Alison,

Thanks for letting me know. Sure I will work towards addressing the
comments as much as possible. Please allow me some time.

Thanks & Regards,
Ashish

On Mon, Nov 17, 2025 at 6:51 PM Alison Kline ***@***.***>
wrote:

> *alsnkline* left a comment (fivetran/fivetran_connector_sdk#335)
> <https://github.com/fivetran/fivetran_connector_sdk/pull/335#issuecomment-3543395145>
>
> Hi @aksaha9 <https://github.com/aksaha9>
> Thank you so much for taking part and submitting your connector to the
> repo - I'm judging right now and would love to see the connector fully
> merged before I finalize my input.
> It is possible for you to complete another pass on the comments?
>
> Thank you so much. - Alison
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/fivetran/fivetran_connector_sdk/pull/335#issuecomment-3543395145>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACELAMJO4IXGJJQH6KSMSOD35IKMFAVCNFSM6AAAAACKGP3PQOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKNBTGM4TKMJUGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

aksaha9 committed:
Changes to address PR review comments

aksaha9 committed:
Further changes to README based on PR feedback

aksaha9 commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,341 @@
+if __name__ == "__main__":
+    with open("configuration.json", "r") as f:
+        configuration = json.load(f)
+    connector.debug(configuration=configuration)
```
Changes made to address all feedback

aksaha9 commented on connectors/owasp_api_vulns/README.md:
```diff
@@ -0,0 +1,111 @@
+  "write_temp_files": "false",
+  "logging_level": "standard",
+  "cwe_ids": "CWE-285,CWE-639,CWE-287,CWE-288,CWE-290,CWE-294,CWE-301,CWE-302,CWE-303,CWE-304,CWE-306,CWE-307,CWE-521,CWE-798,CWE-269,CWE-400,CWE-770,CWE-642,CWE-918,CWE-16,CWE-209,CWE-598,CWE-20,CWE-73,CWE-200,CWE-79"
+}
```
Changes made to address all feedback

aksaha9 committed:
adding W503 to flake8 ignore to allign linter with the formatter

fivetran-sahilkhirwal reviewed (commented):
The changes look good to me
It would be great if the last few copilot comments are addressed
Let me know after the changes are done. After that the PR is good to be merged
Thanks :)

aksaha9 committed:
Changes Addressing the latest PR feedback as well as black and flake8 failures

aksaha9 commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,654 @@
+        except ValueError as e:
+            log.warning(f"Invalid last_sync_time format: {e}; defaulting to full sync")
+            params.pop("lastModStartDate", None)
+            params.pop("lastModEndDate", None)
```
Error handling enhanced to resolve the feedback

aksaha9 commented on connectors/owasp_api_vulns/connector.py:
```diff
@@ -0,0 +1,654 @@
+            last_sync = datetime.datetime.fromisoformat(
+                state["last_sync_time"].replace("Z", "+01:00")
+            )
+            params["lastModStartDate"] = (last_sync - datetime.timedelta(days=1)).strftime(
```
resolved and addressed appropriately

aksaha9 committed:
Changes to connector README to follow template guidelines

aksaha9 committed:
Revert .flake8 config to match upstream main

aksaha9 commented:
Hi @fivetran-sahilkhirwal Request a re review of my latest changes and approve PR if everything ok. Thanks so much

aksaha9 commented on connectors/owasp_api_vulns/README.md:
```diff
@@ -0,0 +1,118 @@
+## Authentication
+This connector authenticates with the NVD 2.0 API using an API key. The key is provided in the `configuration.json` file and sent in the `apiKey` header of each request.
+
+To obtain an API key, you must register on the NVD website.
```
Detailed steps added now

aksaha9 committed:
Changes to address latest Copilot review comments

aksaha9 commented:
Hi Alison ( @alsnkline ),

Thanks so much for assessing, evaluating and judging my connector example. I have tried my best to address all PR comments and also requested another batch of copilot review and addressed them too with the best of my ability. I have requested re-review from reviewers and await their feedback or approval so that I am able to merge. Thanks for your encouragement and inspiration.

Thanks so much.

After the changes I tested in my local with the fivetran debug and the results look good -

(venv) 192:owasp_api_vulns aksaha21$ fivetran debug --configuration configuration_local_test.json | tee local_output_files/run_log_$(date +%Y%m%d_%H%M%S).txt 2>&1
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1763565040.835869 46800009 fork_posix.cc:71] Other threads are currently calling into gRPC, skipping fork() handlers
Nov 19, 2025 03:10:40 PM WARNING: Fivetran-Connector-SDK: `requirements.txt` file not found in your project folder. 
Nov 19, 2025 03:10:40 PM INFO: Fivetran-Connector-SDK: Debugging connector at: /Users/aksaha21/Dropbox/My Mac (Ashishs-MBP)/Documents/mycode/Ashish/Hackathon/Devpost/AI accelerate fivetran/fivetran_connector_sdk/connectors/owasp_api_vulns 
Nov 19, 2025 03:10:40 PM INFO: Fivetran-Connector-SDK: Running connector tester... 
Nov 19, 2025 03:10:41 PM INFO: Fivetran-Tester-Process: Version: 2.25.1031.001 
Nov 19, 2025 03:10:41 PM INFO: Fivetran-Tester-Process: Destination schema: /Users/aksaha21/Dropbox/My Mac (Ashishs-MBP)/Documents/mycode/Ashish/Hackathon/Devpost/AI accelerate fivetran/fivetran_connector_sdk/connectors/owasp_api_vulns/files/warehouse.db/tester 
Nov 19, 2025 03:10:43 PM INFO: Fivetran-Tester-Process: Previous state:
{"cwe_progress": {"CWE-285": "2025-11-19T15:04:07.257500+00:00", "CWE-639": "2025-11-19T15:04:10.350702+00:00", "CWE-287": "2025-11-19T15:04:50.231431+00:00"}} 
Nov 19, 2025 03:10:44 PM INFO: Fivetran-Connector-SDK: Initiating the 'schema' method call... 
Nov 19, 2025 03:10:44 PM INFO: Fivetran-Tester-Process: [SchemaChange]: tester.owasp_api_vulnerabilities 
Nov 19, 2025 03:10:44 PM INFO: Fivetran-Tester-Process: [CreateTable]: tester.owasp_api_vulnerabilities 
Nov 19, 2025 03:10:44 PM INFO: Fivetran-Tester-Process: [SchemaChange]: tester.owasp_api_sync_log 
Nov 19, 2025 03:10:44 PM INFO: Fivetran-Tester-Process: [CreateTable]: tester.owasp_api_sync_log 
Nov 19, 2025 03:10:44 PM INFO: Fivetran-Connector-SDK: Initiating the 'update' method call... 
Nov 19, 2025 03:10:44 PM WARNING: Example: CONNECTOR: OWASP_API_VULNS 
Nov 19, 2025 03:10:44 PM INFO: Now pulling CVEs pertaining to CWE-285 
Nov 19, 2025 03:10:45 PM INFO: API response for CWE-285: status=200, totalResults=272, startIndex=0 
Nov 19, 2025 03:10:45 PM INFO: Results fetched for CWE-285 
Nov 19, 2025 03:10:45 PM INFO: Rows upserted for CWE-285: 215 
Nov 19, 2025 03:10:45 PM INFO: Now pulling CVEs pertaining to CWE-639 
Nov 19, 2025 03:10:47 PM INFO: Fivetran-Tester-Process: Checkpoint: {"cwe_progress": {"CWE-285": "2025-11-19T15:10:45.026202+00:00"}} 
Nov 19, 2025 03:10:49 PM INFO: API response for CWE-639: status=200, totalResults=883, startIndex=0 
Nov 19, 2025 03:10:49 PM INFO: Results fetched for CWE-639 
Nov 19, 2025 03:10:49 PM INFO: Rows upserted for CWE-639: 791 
Nov 19, 2025 03:10:49 PM INFO: Now pulling CVEs pertaining to CWE-287 
Nov 19, 2025 03:10:50 PM INFO: API response for CWE-287: status=200, totalResults=3318, startIndex=0 
Nov 19, 2025 03:10:51 PM INFO: API response for CWE-287: status=200, totalResults=3318, startIndex=2000 
Nov 19, 2025 03:10:58 PM INFO: Fivetran-Tester-Process: Checkpoint: {"cwe_progress": {"CWE-285": "2025-11-19T15:10:45.026202+00:00", "CWE-639": "2025-11-19T15:10:49.156230+00:00"}} 
Nov 19, 2025 03:11:04 PM INFO: Results fetched for CWE-287 
Nov 19, 2025 03:11:04 PM INFO: Rows upserted for CWE-287: 1709 
Nov 19, 2025 03:11:04 PM INFO: Sync completed. Total duration: 20.00 seconds 
Nov 19, 2025 03:11:04 PM INFO: Total upserts across all CWEs: 2715 
Nov 19, 2025 03:11:15 PM INFO: Fivetran-Tester-Process: Checkpoint: {"cwe_progress": {"CWE-285": "2025-11-19T15:10:45.026202+00:00", "CWE-639": "2025-11-19T15:10:49.156230+00:00", "CWE-287": "2025-11-19T15:11:04.432844+00:00"}} 
Nov 19, 2025 03:11:15 PM INFO: Fivetran-Tester-Process: SYNC PROGRESS:
Operation       | Calls     
----------------+------------
Upserts         | 2716      
Updates         | 0         
Deletes         | 0         
Truncates       | 0         
SchemaChanges   | 2         
Checkpoints     | 3         
Note: Fivetran debug's performance is limited by your local machine's resources. Your connector will run faster in production.
read about production system resources at https://fivetran.com/docs/connector-sdk/working-with-connector-sdk#systemresources 

Regards,
Ashish

fivetran-pranavtotala reviewed (dismissed):
LGTM

fivetran-dejantucakov committed:
Revise README for clarity and formatting improvements

Updated contributor section and improved formatting for clarity. Revised API key instructions and added descriptions for created tables.

fivetran-dejantucakov commented on connectors/owasp_api_vulns/README.md:
```diff
@@ -0,0 +1,133 @@
+- `force_full_sync` (optional) - Set to `true` to ignore the saved state and perform a full re-sync. Defaults to `false`.
+- `write_temp_files` (optional) - Set to `true` to save the raw API responses to a local `raw_data` directory for debugging. Defaults to `false`.
+- `logging_level` (optional) - Set to `standard` for summary logging or `debug` for verbose, detailed logging. Defaults to `debug`.
+- `cwe_ids` (optional) - A comma-separated string of CWE IDs to filter the vulnerabilities. Defaults to a predefined list of OWASP-related CWEs.
```
@aksaha9 Can you please check the updates here (what is required and what is optional)? TIA

fivetran-dejantucakov replied:
Thank you, approving the PR!

fivetran-dejantucakov reviewed (approved):
@aksaha9 Docs LGTM, thank you!

aksaha9 commented on connectors/owasp_api_vulns/configuration.json.example:
```diff
@@ -0,0 +1,7 @@
+  "api_key": "<YOUR_NVD_API_KEY>",
+  "force_full_sync": "<ENABLE_FORCE_FULL_SYNC>",
+  "write_temp_files": "<ENABLE_TEMP_FILE_WRITE>",
+  "logging_level": "<YOUR_LOGGING_LEVEL>",
```
Placeholder is appropriate here. The possible values are described in the connector README.md

aksaha9 commented:
Hi @fivetran-sahilkhirwal and @fivetran-pranavtotala - Request your re-review and approval. Thanks, Ashish

fivetran-dejantucakov commented:
> Hi @fivetran-sahilkhirwal and @fivetran-dejantucakov , Just a gentle reminder to help me with the merge as I do not have the merge permissions. If either of you could handle the merge for me that would be great. TIA

@aksaha9 Thanks for the reminder, PR has been merged. Thank you for your contribution!
